### ISSUE TITLE
Check RO X-1 Cockpit

### ISSUE THREAD

**Issue by JonnyOThan** [harveyt/reviva#28]
https://forum.kerbalspaceprogram.com/topic/206744-112x-reviva-the-iva-revival-and-editorflight-switcher-bug-fixing-release-100-17th-feb-2024/?do=findComment&comment=4374998

**Comment by PubliusKerman** [comment]
Too big to attach all the logs in zip, drop box download at link.

**Comment by PubliusKerman** [comment]
I've been doing some digging and it seems related to how they added their x-15 and x-1 cockpits into the game by cloning the the Mk1 and Mk1inline and then swapping the cockpit internals and giving the mk1CockpitInternal to their X-1(RO-x1cockpit, a modified AP+ x1cockpit), example below but you can find the rest in the SuggestedMods mods folder of RO, got a whole folder on Squad hidden in there. I've also noticed a pattern, it seems like any of the MAS MFD's or HUD works(think it's the MAS one, the one in DE and RPM but not ASET or KSA) Including the MAS MFD in the "DE + MAS" cockpits which along with the HUD still work while nothing else does.

+PART[Mark1Cockpit]:FOR[RealismOverhaul]
{
    @name = RO-Mk1Cockpit
    @title = Mk1 Cockpit
    @description = Two-person spaceplane cockpit. Rated for LEO reentries.
    @mass = 1.3
    @INTERNAL
    {
        @name = mk2CockpitStandardInternals
        %offset = 0.0, -0.2, -0.9
    }
}

**Comment by JonnyOThan** [comment]
[RO - full iva - debug logs (2).zip](https://github.com/harveyt/reviva/files/14747076/RO.-.full.iva.-.debug.logs.2.zip)

Looks like you didn't enable compression on the zip file.

**Comment by PubliusKerman** [comment]
lol, that's dropbox for you. I just pulled it straight from them and just assumed cause it came in a zip when I tested link.

**Comment by PubliusKerman** [comment]
I was going through the Reviva MM patches and I think there might be an easy solution by modifying the backstop/catch-all. Here is an example of the change.

[FlightComputers.txt](https://github.com/harveyt/reviva/files/14754297/FlightComputers.txt)
[FlightComputers - modified.txt](https://github.com/harveyt/reviva/files/14754306/FlightComputers.-.modified.txt)

**Comment by PubliusKerman** [comment]
I was doing some digging in the Reviva MM files and think I may have found an elegant solution. Modified MM file below, testing now, results to follow

Reviva/2_Backstop/
[FlightComputers - modified.txt](https://github.com/harveyt/reviva/files/14755546/FlightComputers.-.modified.txt)

**Comment by PubliusKerman** [comment]
It works! I'm going to continue working on a more tailored solution for RO/RP-1, but if you want to push it out, I think it would be a good addition to your backstop for any mods that use cloned parts.

**Comment by JonnyOThan** [comment]
> I was doing some digging in the Reviva MM files and think I may have found an elegant solution. Modified MM file below, testing now, results to follow
> 
> Reviva/2_Backstop/ [FlightComputers - modified.txt](https://github.com/harveyt/reviva/files/14755546/FlightComputers.-.modified.txt)

Can you elaborate on what the actual problem was and how this fixes it?  At first glance, it looks like it would add flight computer to things that might not even have internals which wouldn’t be good.

**Comment by PubliusKerman** [comment]
They are just missing either RPM or MAS computers because of the part cloning I was telling you about. I just figured if it had a command module than it would have an internal, but that's because I use probecontrolroom so much it didn't even occur to me that there's a slight chance it might cause problems. I'm sure there are rare cases where that could cause problems. More tailored MM file below, now referencing INTERNAL instead of CommandModule.

GameData/Reviva/2_Backstop/
[FlightComputers - tailored.txt](https://github.com/harveyt/reviva/files/14758227/FlightComputers.-.tailored.txt)

**Comment by JonnyOThan** [comment]
I'm still not sure if Reviva is the right place for this patch but I'll have to look into the details further.  Like, isn't there still a problem if you don't have reviva installed?

**Comment by PubliusKerman** [comment]
I'll check and see if problem persists without Reviva, never actually tried it. If you one have one IVA installed or none that conflict then it should go to right place because RO was quite elegant about getting new parts in without actually adding part files and adding to bloat. They cloned the mk1 cockpit and gave it a new name and used a Textures Unlimited to edit the entire mesh and texture to make it the X-15, and then slapped the mk2 cockpit in the original, scaled it up to the mk2 size and gave it two seats.

**Comment by PubliusKerman** [comment]
Yes, the ASET cockpits work without Reviva installed. This is because Reviva is trying to make everything run on MAS and if it can't adds original and blended by essentially, but not literally, breaking down all the cockpits and then puts them back together in their cockpits in original, blended RPM and MAS in the case of DE_IVA, or MAS only if available or RPM is not required like KSA IVA.

**Comment by JonnyOThan** [comment]
> I'll check and see if problem persists without Reviva, never actually tried it. If you only have one IVA installed or none that conflict then it should go to right place because RO was quite elegant about getting new parts in without actually adding part files and memory bloat. They cloned the mk1 cockpit and gave it a new name and used Textures Unlimited to edit the entire mesh and texture to make it the X-15, and then slapped the mk2 cockpit in the original, scaled it up to the mk2 size and gave it two seats.

That sounds fragile not elegant.

**Comment by PubliusKerman** [comment]
That's because it isn't meant to be elegant like the MM patch for RO I'm working on now, it's meant to be brute like your backstop so it can also catch things from cloned parts. When someone uses MM to clone a part and then modify which internal it uses, Reviva has no way of handling this and at least placing the necessary flight computer MODULES for the cockpits to function, which is why half the MFDs didn't work unless they were all MAS. If you don't have Reviva installed and another mod clones a stock part through MM and then modifies which cockpit is has the IVA mod just goes and finds any part with that internal and then modifies from there, which is why they work when Reviva is not installed.

**Comment by PubliusKerman** [comment]
The reason I provided the modified backstop for submission is just in case another mod uses the same solution in the future or RO team decides to clone another part. The modified backstop will now catch them as well so that Reviva doesn't completely break all their RPM or MAS IVAs. Work continues on the tailored RO patch, will update when I have something.
